Initial Sync Health Status Class and Data Loss Handler

roccotripaldi opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
* Adds a new class to manage sync health
* Adds a new Jetpack_Option to track sync health
* Adds unit tests that outline how the class is used throughout the code base

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This adds a new concept of Sync Health and API to properly determine if the Jetpack site is properly syncing data to WordPress.com
* Discussed in detail here: p9dueE-1bp-p2

#### Testing instructions:
- If the unit tests are passing, and @mdbitz thinks the logic makes sense, this should be good to go
- It's difficult to test this on an actual site at the moment, because Sync Health is such a new concept, and we don't have any tools to help us fake an out of sync site.
- This work is part of a bigger sprint, and we'll do more in-depth QA in the next PRs which should be landing later this week

mdbitz committed:
[not verified] Initial Sync Health Status Class and Data Loss Handler

mdbitz committed:
[not verified] Initial Sync Health Status Class and Data Loss Handler

roccotripaldi committed:
Renaming stuff for clarity.

mdbitz committed:
Merge branch 'master' of github.com:Automattic/jetpack into add/sync-health

roccotripaldi committed:
Ensure we use `Jetpack_Options` to manage options.

mdbitz committed:
[not verified] Add Full Sync End handler to update Status

mdbitz committed:
[not verified] Merge Rocco's changes to normalize to JP Options

mdbitz committed:
[not verified] First argument is checksum that was deprecated

mdbitz committed:
[not verified] Activation and upgrade hooks to set Status to Unknown

mdbitz committed:
[not verified] Initializing Status to allow us to track if this is initial setup.

mdbitz commented on class.jetpack.php:
```diff
@@ -3162,6 +3162,11 @@ public static function plugin_activation( $network_wide ) {
 
 		update_option( 'jetpack_activation_source', self::get_activation_source( wp_get_referer() ) );
 
+		// Define Status as "Unknown"
```
"Initializing"  sorry last commit changed the code on 3167 but not here in the comment

roccotripaldi committed:
Fixing spacing and comments.

roccotripaldi committed:
Jetpack Sync: Renaming Status to Health

We already have a Jetpack Status class, and we have a sync-status endpoint.
Sync Health is less ambiguous.

roccotripaldi committed:
[not verified] Jetpack Sync: adding base tests for Sync Health

Not verified because most of our linter rules should not apply to
unit test.

roccotripaldi committed:
[not verified] Jetpack Sync: actions unit tests

Adding Health-related unit tests to Sync Actions tests.
Not verified because our linting rules should not apply to unit tests.

roccotripaldi committed:
[not verified] Jetpack Sync: unit tests

add Health-related unit tests for Full Sync and Sync Listener.
not verified because our linter rules should not be applied to unit tests.

roccotripaldi committed:
Sync: actions

Only perform data loss routine if sync is enabled.

roccotripaldi committed:
Sync Health: disabled state

Account for Jetpack Sync's health when sync is disabled.

roccotripaldi committed:
[not verified] Sync Health: unit tests

Adds unit tests for Health's disabled state.
Not verified because linter logic should not be applied to unit tests.

roccotripaldi committed:
Sync

Data loss routine now checks for an enabled sync within it's method
rather than within the enqueue_action method

It is important that enqueue_action logic remains intact.
I'm grateful that a unit test caught it.

mdbitz commented:
woah look at all these changes. I'll review in the morning before diving back into es queries :)

mdbitz reviewed (commented):
@roccotripaldi  Really like where this is headed and the tests 😱 Horray for Unit Tests.

I left some feedback on a few of the state changes. Let me know your thoughts as I could see some use cases that might complicate things.

roccotripaldi committed:
Sync: Health

Move full-sync hooks out of actions::init to health::init

roccotripaldi committed:
[not verified] Sync: Health

Adjust logic to be more specific. Plugin upgrade and plugin activation will have their own methods.
- Not verified because linting rules don't make sense for phpunit tests.

roccotripaldi committed:
Sync Health

No need for an initialization status.

roccotripaldi commented:
@mdbitz - thanks for the review. I addressed all your feedback.

I agree with the logic changes you suggested, and because of that, we currently have no need for `STATUS_INITIALIZING` because it will be difficult to determine this during activation and upgrade hooks.

Perhaps we can ID a better place to set `STATUS_INITIALIZING` - but i'd like to do that in an other PR. 

I prefer to not keep PRs open for more than 24 hours, and to keep them fairly small.

mdbitz reviewed (approved):
Thanks for those updates, Rocco. I agree on initializing let's keep that separate for now and we can PR if we come across a specific use case that needs it.

Tests are green and local testing is showing expected behavior.

mdbitz commented on packages/sync/src/class-listener.php:
```diff
@@ -353,6 +356,28 @@ public function enqueue_action( $current_filter, $args, $queue ) {
+		if ( ! Settings::is_sync_enabled() ) {
+			return;
+		}
+		Health::update_status( Health::STATUS_OUT_OF_SYNC );
```
Just thought of an edge case. This will give us a large number of 'jetpack_sync_data_loss' events potentially if the site keeps trying to enqueue. We should check if status is already out_of_sync. If so then don't update or send the action.

basically add an || Health::STATUS_OUT_OF_SYNC === Health::get_status() to line 366
